### PR TITLE
Support for composable functions

### LINKED ISSUE #402: /range/lastCell is missing in OpenAPI description
Path:
`/me/drive/items/{id}/workbook/names/{name}/range/lastCell` is missing in OpenAPI description.

Raptor error:
Original HTTP Snippet:
GET /v1.0/me/drive/items/{id}/workbook/names/{name}/range/lastCell HTTP/1.1
Host: graph.microsoft.com

Exception while processing /home/vsts/work/1/a/Snippets/range-lastcell-v1-httpSnippet.
Path segment 'lastCell' not found in path.

More of these reported here: https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/1873

Originally reported: https://github.com/microsoftgraph/msgraph-metadata/issues/385

### LINKED ISSUE #404: /visibleView/rows missing in OpenAPI description
Path:
`/v1.0/me/drive/root/workbook/worksheets/{id}/range(address='A1:Z10')/visibleView/rows` is missing in the openAPI descrition.

Raptor error:
Original HTTP Snippet:
GET /v1.0/me/drive/root/workbook/worksheets/{id}/range(address='A1:Z10')/visibleView/rows HTTP/1.1
Host: graph.microsoft.com

Exception while processing /home/vsts/work/1/a/Snippets/get-rows-v1-httpSnippet.
Path segment 'visibleView' not found in path

We checked in the openAPI file and confirmed the path is not defined.

Originally reported here: https://github.com/microsoftgraph/msgraph-metadata/issues/384

### PR DESCRIPTION
**Pull request by MartinM85**
fixes #402
fixes #404

Generate ODataPath for composable functions

### CODE REVIEW

**Comment by MartinM85** [comment]
Proposal of a fix for #402.

Now, `ODataPath` is not generated for composable functions

`../operation/operation`

Example from Graph API v1.0 metadata.xml

```
      <Function Name="lastCell" IsBound="true" IsComposable="true">
        <Parameter Name="bindparameter" Type="graph.workbookRange" />
        <ReturnType Type="graph.workbookRange" />
      </Function>

      <Function Name="visibleView" IsBound="true" IsComposable="true">
        <Parameter Name="bindparameter" Type="graph.workbookRange" />
        <ReturnType Type="graph.workbookRangeView" />
      </Function>

      <Action Name="unmerge" IsBound="true">
        <Parameter Name="bindparameter" Type="graph.workbookRange" />
      </Action>
```

If functions are composable, they can be chained 

`../operation/operation/../operation`

I have a restriction that only **one** operation segment can be appended

It's not a fix for #404 which is a case when a navigation segment is added to the operation segment

I can add the navigation segment

```
...
AppendPath(newOperationPath);

foreach (var navProperty in bindingEntityType.NavigationProperties())
{
    ODataPath newNavigationPath = newOperationPath.Clone();
    newNavigationPath.Push(new ODataNavigationPropertySegment(navProperty));
    AppendPath(newNavigationPath);
}
```

but `newNavigationPath.Kind` is calculated as `Operation` and it will break the code in https://github.com/microsoft/OpenAPI.NET.OData/blob/aaa82b320fd99f5d93a716a032f72c01374bee07/src/Microsoft.OpenApi.OData.Reader/PathItem/OperationPathItemHandler.cs#L75

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 1]
```diff
@@ -708,7 +708,10 @@ private void CreateTypeCastPaths(ODataPath currentPath, OpenApiConvertSettings c
         /// </summary>
         private void RetrieveBoundOperationPaths(OpenApiConvertSettings convertSettings)
         {
-            foreach (var edmOperation in _model.GetAllElements().OfType<IEdmOperation>().Where(e => e.IsBound))
+            // process functions before actions
+            var edmFunctions = _model.GetAllElements().OfType<IEdmFunction>().Where(e => e.IsBound).ToList();
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:712]:
```suggestion
            var edmFunctions = _model.GetAllElements().OfType<IEdmFunction>().Where(e => e.IsBound);
```

you're only enumerating once, no need for the to list

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 2]
```diff
@@ -708,7 +708,10 @@ private void CreateTypeCastPaths(ODataPath currentPath, OpenApiConvertSettings c
...
+            var edmActions = _model.GetAllElements().OfType<IEdmOperation>().Where(e => e.IsBound && e is not IEdmFunction).ToList();
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:713]:
```suggestion
            var edmActions = _model.GetAllElements().OfType<IEdmAction>().Where(e => e.IsBound);
```

Could you not do that instead?

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 3]
```diff
@@ -993,5 +999,54 @@ private void AppendBoundOperationOnNavigationPropertyPath(IEdmOperation edmOpera
...
         }
+
+        private void AppendBoundOperationOnOperationPath(IEdmOperation edmOperation, bool isCollection, IEdmEntityType bindingEntityType)
+        {
+            bool isEscapedFunction = _model.IsUrlEscapeFunction(edmOperation);
+
+            // only composable functions
+            var paths = _allOperationPaths.Where(x => x.LastSegment is ODataOperationSegment operationSegment
+                                                    && operationSegment.Operation is IEdmFunction edmFunction
+                                                    && edmFunction.IsComposable).ToList();
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1010]:
```suggestion
                                                    && edmFunction.IsComposable);
```

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 4]
```diff
@@ -993,5 +999,54 @@ private void AppendBoundOperationOnNavigationPropertyPath(IEdmOperation edmOpera
...
+
+            foreach (var path in paths)
+            {
+                if (path.LastSegment is ODataOperationSegment operationSegment)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1014]: negate this statement and continue when true to reduce the arrow head style of the code here

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 5]
```diff
@@ -993,5 +999,54 @@ private void AppendBoundOperationOnNavigationPropertyPath(IEdmOperation edmOpera
...
+                {
+                    if (path.Segments.Count > 1 && path.Segments[path.Segments.Count - 2] is ODataOperationSegment)
+                    {
+                        continue;
+                    }
+
+                    if (operationSegment.Operation is IEdmFunction edmFunction && edmFunction.IsComposable)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1021]: negate and continue

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 6]
```diff
@@ -993,5 +999,54 @@ private void AppendBoundOperationOnNavigationPropertyPath(IEdmOperation edmOpera
...
+                    {
+                        if (edmFunction.ReturnType != null && edmFunction.ReturnType.Definition.Equals(bindingEntityType))
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1023]: negate and continue

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 7]
```diff
@@ -993,5 +999,54 @@ private void AppendBoundOperationOnNavigationPropertyPath(IEdmOperation edmOpera
...
+                        {
+                            if (!isCollection)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1025]: remove negation and continue

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for the great contribution! In addition to my comments, could you add unit tests for the case you're implementing please?

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 3 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1010]: `AppendPath` inside the `foreach` adds a new path to `_allOperationPaths`.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 1 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: Reverted.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 2 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: Reverted

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 4 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: fixed

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 5 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: fixed

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 6 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: fixed

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 7 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: fixed

**Comment by MartinM85** [comment]
Added unit test and small changes to ensure that operations are appended to all composable functions, and navigation properties are appended to functions.

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 8]
```diff
@@ -757,7 +745,81 @@ private void RetrieveBoundOperationPaths(OpenApiConvertSettings convertSettings)
                     AppendBoundOperationOnDerivedNavigationPropertyPath(edmOperation, isCollection, bindingEntityType, convertSettings);
                 }
             }
+
+            // all operations appended to properties
+            // append bound operations to functions
+            foreach (var edmOperation in _model.GetAllElements().OfType<IEdmOperation>().Where(x => x.IsBound))
```
[the remark this answers is not in the document]
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: Changed

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 9]
```diff
@@ -757,7 +745,81 @@ private void RetrieveBoundOperationPaths(OpenApiConvertSettings convertSettings)
...
+                    AppendBoundOperationOnOperationPath(edmOperation, isCollection, bindingEntityType);
+                }
+            }
+
+            // append navigation properties to functions with return type
+            var functionPaths = _allOperationPaths.Where(x => x.LastSegment is ODataOperationSegment operationSegment
+                                                    && operationSegment.Operation is IEdmFunction edmFunction
+                                                    && edmFunction.IsComposable
+                                                    && edmFunction.ReturnType != null
+                                                    && edmFunction.ReturnType.Definition is IEdmEntityType returnBindingEntityType).ToList();
```
[the remark this answers is not in the document]
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs]: fixed

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 3 above)
Comment (MartinM85) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:1067]: yes, but in fact, `_allOperationPaths` will be enumerated and an attempt to add a new operation path to `_allOperationPaths` will throw `InvalidOperationException : Collection was modified; enumeration operation may not execute.`

**Review by baywet — DISMISSED** [review]
Thanks for making the changes! @irvinesunday for final review and merge

**Comment by irvinesunday** [comment]
Wanted to add this diff screenshot for informational purposes.

image

The diff image above shows the current number of paths per API generated using the [Kiota-based CSDL](https://github.com/microsoftgraph/msgraph-metadata/blob/master/clean_v10_metadata/cleanMetadataWithDescriptionsAndAnnotationsAndErrorsv1.0.xml) for v1.0.
**Left** is current production figures, **right** is with the changes in this PR.

Below is a similar diff but using the [non Kiota-based CSDL](https://github.com/microsoftgraph/msgraph-metadata/blob/master/clean_v10_metadata/cleanMetadataWithDescriptionsAndAnnotationsv1.0.xml) (PowerShell and Graph Explorer)

image

**Left** is current production figures, **right** is with the changes in this PR.

**drives** API seem to have grown by quite a huge margin. We may need to investigate the accuracy of the operations bound to the drives API and and fix them in the metadata if necessary before merging this PR.

**Comment by irvinesunday** [comment]
@MartinM85 do you mind updating the release notes and bumping up the version in the csproj file?

**Comment by MartinM85** [comment]
@irvinesunday Version and release notes updated

**Comment by irvinesunday** [comment]
> Wanted to add this diff screenshot for informational purposes.
> 
> image
> 
> The diff image above shows the current number of paths per API generated using the [Kiota-based CSDL](https://github.com/microsoftgraph/msgraph-metadata/blob/master/clean_v10_metadata/cleanMetadataWithDescriptionsAndAnnotationsAndErrorsv1.0.xml) for v1.0. **Left** is current production figures, **right** is with the changes in this PR.
> 
> Below is a similar diff but using the [non Kiota-based CSDL](https://github.com/microsoftgraph/msgraph-metadata/blob/master/clean_v10_metadata/cleanMetadataWithDescriptionsAndAnnotationsv1.0.xml) (PowerShell and Graph Explorer)
> 
> image
> 
> **Left** is current production figures, **right** is with the changes in this PR.
> 
> **drives** API seem to have grown by quite a huge margin. We may need to investigate the accuracy of the operations bound to the drives API and and fix them in the metadata if necessary before merging this PR.

Investigation to the growth in the `drives` API is tracked with this [issue](https://github.com/microsoft/OpenAPI.NET.OData/issues/439).

**Comment by irvinesunday** [comment]
Tested this again, the changes are acceptable, and this PR can be merged.

image
_v1.0_

**Comment by baywet** [comment]
@irvinesunday, thanks for double checking. Did you have to make any changes to the metadata? Any changes we should make to the test CSDLs here considering the bigger increase in path items? (see the integration tests numbers)

**Comment by irvinesunday** [comment]
> @irvinesunday, thanks for double checking. Did you have to make any changes to the metadata? Any changes we should make to the test CSDLs here considering the bigger increase in path items? (see the integration tests numbers)

No, I didn't make any changes to the metadata. The integration file is probably falling behind the main metadata file in terms of annotations.

**Comment by baywet** [comment]
@irvinesunday would you mind having a quick look at those to check if something obvious comes up? (like a wrong contains target or a missing annotation) so we can keep those numbers down?

**Review by baywet — APPROVED** [review]
Thanks everyone for the work on this one.
@irvinesunday for final review and merge.

**Comment by irvinesunday** [comment]
@baywet our biggest culprit was a containment for a `directoryObject` type navigation property that also had a collection of derived types. The integration file and our prod file are still not 100% a match but nothing else out of the ordinary stood out.

```XML
<NavigationProperty Name="deletedItems" Type="Collection(graph.directoryObject)" ContainsTarget="true">
  <Annotation Term="Org.OData.Validation.V1.DerivedTypeConstraint">
    <Collection>
      <String>microsoft.graph.user</String>
      <String>microsoft.graph.group</String>
      <String>microsoft.graph.application</String>
    </Collection>
  </Annotation>
  <Annotation Term="Org.OData.Core.V1.ExplicitOperationBindings">
    <Collection>
      <String>microsoft.graph.restore</String>
    </Collection>
  </Annotation>
</NavigationProperty>
```
Removing the containment (as we've done in our prod files) lowers the path numbers here significantly.
We can merge this.

**Review by irvinesunday — APPROVED** [review]

**Pull request merged**